cmd/restic/cmd_prune.go

```diff
@@ -82,226 +123,400 @@ func runPrune(gopts GlobalOptions) error {
+			bar.Start()
+			// TODO in Repack:  - Parallelize repacking
+			//                  - Save full indexes during repacking
+			//                  - Make use of blobs stored multiple times (e.g. if SHA doesn't match)
+			_, err := repository.Repack(ctx, repo, repackPacks, keepBlobs, bar)
+			bar.Done()
+			if err != nil {
+				return err
+			}
+		} else {
+			if !gopts.JSON {
+				for id := range repackPacks {
+					Verbosef("would have repacked data file %v.\n", id.Str())
+				}
+			}
+		}
+
+		// Also remove repacked packs
+		removePacks.Merge(repackPacks)
 	}
 
-	if len(removePacks) != 0 {
-		bar = newProgressMax(!gopts.Quiet, uint64(len(removePacks)), "packs deleted")
-		bar.Start()
-		for packID := range removePacks {
-			h := restic.Handle{Type: restic.DataFile, Name: packID.String()}
-			err = repo.Backend().Remove(ctx, h)
+	if len(removePacks) != 0 && !opts.DryRun {
+		Verbosef("updating index files...\n")
+
+		if opts.NoRebuildIndex {
+			// Call RebuildIndex: rebuilds the index from the already loaded in-memory index.
+			// TODO in RebuildIndex: - Save full indexes
+			//						- Parallelize repacking
+			newIndex, obsoleteIndexes, err := (repo.Index()).(*repository.MasterIndex).RebuildIndex(removePacks)
 			if err != nil {
-				Warnf("unable to remove file %v from the repository\n", packID.Str())
+				return err
+			}
+
+			_, err = repository.SaveIndex(ctx, repo, newIndex)
+			if err != nil {
+				return err
+			}
+
+			Verbosef("deleting obsolete index files...\n")
+			err = DeleteFiles(gopts, opts.DryRun, false, repo, obsoleteIndexes, restic.IndexFile)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Call "restic rebuild-index": rebuild the index from pack files
+			if err = rebuildIndex(ctx, repo, removePacks); err != nil {
+				return err
 			}
-			bar.Report(restic.Stat{Blobs: 1})
 		}
-		bar.Done()
 	}
 
-	Verbosef("done\n")
+	if len(removePacks) != 0 {
+		Verbosef("deleting obsolete data files...\n")
+		DeleteFiles(gopts, opts.DryRun, true, repo, removePacks, restic.DataFile)
+	}
+
+	Verbosef("done.\n")
 	return nil
 }
+
+const numDeleteWorkers = 8
+
+// DeleteFiles deletes the given fileList of fileType in parallel
+// if dryrun=true, it will just print out what would be deleted
+// if ignoreError=true, it will print a warning if there was an error, else it will abort.
+func DeleteFiles(gopts GlobalOptions, dryrun bool, ignoreError bool, repo restic.Repository, fileList restic.IDSet, fileType restic.FileType) error {
+	totalCount := len(fileList)
+	fileHandles := make(chan restic.Handle)
+	go func() {
+		for id := range fileList {
+			fileHandles <- restic.Handle{Type: fileType, Name: id.String()}
+		}
+		close(fileHandles)
+	}()
+
+	bar := newProgressMax(!gopts.Quiet, uint64(totalCount), "files deleted")
+	wg, ctx := errgroup.WithContext(gopts.ctx)
+	bar.Start()
+	for i := 0; i < numDeleteWorkers; i++ {
+		wg.Go(func() error {
+			for h := range fileHandles {
+				if !dryrun {
```

Comment:
Same here: the dry run case should be treated specially, the regular case unindented. This one is trickier because the bar.Report call should happen if (dryrun || err == nil), but it's short enough to just repeat:

```
if dryrun {
	if !gopts.JSON {
		Verbosef("would have removed %v.\n", h.Name)
	}
	bar.Report(restic.Stat{Blobs: 1})
	continue
}

// former !dryrun case here
bar.Report(restic.Stat{Blobs: 1})

Reply:
is changed.